src/OrchardCore/OrchardCore.Navigation.Core/NavigationProviderBase.cs

```diff
@@ -0,0 +1,38 @@
+namespace OrchardCore.Navigation;
+
+public abstract class NavigationProviderBase : INavigationProvider
+{
+    /// <summary>
+    /// The name of the navigation menu.
+    /// </summary>
+    protected abstract string Name { get; }
+
+    public ValueTask BuildNavigationAsync(string name, NavigationBuilder builder)
+    {
+        if (name != Name)
+        {
+            return ValueTask.CompletedTask;
+        }
+
+        return BuildAsync(builder);
+    }
+
+    /// <summary>
+    /// Asynchronously builds the navigations menu.
+    /// </summary>
+    /// <param name="builder"></param>
+    protected virtual ValueTask BuildAsync(NavigationBuilder builder)
+    {
+        Build(builder);
+
+        return ValueTask.CompletedTask;
+    }
+
+    /// <summary>
+    /// Synchronously builds the navigations menu.
+    /// </summary>
+    /// <param name="builder"></param>
+    protected virtual void Build(NavigationBuilder builder)
```

Comment:
The only advantage of `Build` instead of `BuildAsync` is that you can return nothing instead of `ValueTask.CompletedTask`.

With only adding a filter on `name` I don't think this PR is a good thing since it's adding new concepts to discover/understand (NavigationProviderBase vs AdminNavigationProvider) where before anything would implement INavigationProvider and check the name.

Also adding a breaking change (IsAdminMenu) though this can be mitigated with an [Obsolete].

Reply:
@sebastienros 

> Also adding a breaking change (IsAdminMenu) though this can be mitigated with an [Obsolete].

This isn't a breaking change because `IsAdminMenu` was added in 2.0 preview. So removing it now is not a breaking change. But if we don't include this PR in 2.0, then it will become a breaking change unless we use `[Obsolete]` to not make it breaking.

The main key here is to ask the users to make all the changes to the AdminMenu at once.

> With only adding a filter on name I don't think this PR is a good thing since it'

People don't have to follow our clean code approach and use this new base class, they can continue to write a longer code if they want to. With this approach it is just a much cleaner way to implement an navigation in out admin menu.